Mathlib/Algebra/Category/ModuleCat/Differentials/Basic.lean

```diff
@@ -0,0 +1,185 @@
+
+@[simp]
+lemma d_mul (b b' : B) : D.d (b * b') = b • D.d b' + b' • D.d b := by simp [d]
+
+@[simp]
+lemma d_map (a : A) : D.d (f a) = 0 :=
+  letI := f.toAlgebra
+  letI := Module.compHom M f
+  D.map_algebraMap a
+
+end Derivation
+
+end ModuleCat
+
+namespace CommRingCat
+
+variable {A B A' B' : CommRingCat.{u}} {f : A ⟶ B} {f' : A' ⟶ B'}
+  {g : A ⟶ A'} {g' : B ⟶ B'} (fac : g ≫ f' = f ≫ g')
+
+variable (f) in
+/-- The module of differentials of a morphism `f : A ⟶ B` in the category `CommRingCat`. -/
+noncomputable def KaehlerDifferential : ModuleCat.{u} B :=
+  letI := f.toAlgebra
+  ModuleCat.of B (_root_.KaehlerDifferential A B)
+
+namespace KaehlerDifferential
+
+variable (f) in
+/-- The (universal) derivation in `(KaehlerDifferential f).Derivation f`
+when `f : A ⟶ B` is a morphism in the category `CommRingCat`. -/
+noncomputable def D : (KaehlerDifferential f).Derivation f :=
+  letI := f.toAlgebra
+  ModuleCat.Derivation.mk
+    (fun b ↦ _root_.KaehlerDifferential.D A B b) (by simp) (by simp)
+      (_root_.KaehlerDifferential.D A B).map_algebraMap
+
+/-- When `f : A ⟶ B` is a morphism in the category `CommRingCat`, this is the
+differential map `B → KaehlerDifferential f`. -/
+noncomputable abbrev d (b : B) : KaehlerDifferential f := (D f).d b
+
+@[ext]
+lemma ext {M : ModuleCat B} {α β : KaehlerDifferential f ⟶ M}
+    (h : ∀ (b : B), α (d b) = β (d b)) : α = β := by
+  rw [← sub_eq_zero]
+  have : ⊤ ≤ LinearMap.ker (α - β) := by
+    rw [← KaehlerDifferential.span_range_derivation, Submodule.span_le]
+    rintro _ ⟨y, rfl⟩
+    rw [SetLike.mem_coe, LinearMap.mem_ker, LinearMap.sub_apply, sub_eq_zero]
+    apply h
+  rw [top_le_iff, LinearMap.ker_eq_top] at this
+  exact this
+
+/-- The map `KaehlerDifferential f ⟶ (ModuleCat.restrictScalars g').obj (KaehlerDifferential f')`
+induced by a commutative square (given by an equality `g ≫ f' = f ≫ g'`)
+in the category `CommRingCat`. -/
+noncomputable def map :
+    KaehlerDifferential f ⟶
+      (ModuleCat.restrictScalars g').obj (KaehlerDifferential f') :=
+  letI := f.toAlgebra
+  letI := f'.toAlgebra
+  letI := g.toAlgebra
+  letI := g'.toAlgebra
+  letI := (g ≫ f').toAlgebra
+  have := RingHom.isScalarTower_toAlgebra_comp g f' _ rfl
+  have := RingHom.isScalarTower_toAlgebra_comp f g' _ fac.symm
+  have := RingHom.smulCommClass_toAlgebra g' f'
+  { toFun := fun x ↦ _root_.KaehlerDifferential.map A A' B B' x
+    map_add' := by simp
+    map_smul' := by simp }
+
+@[simp]
+lemma map_d (b : B) : map fac (d b) = d (g' b) := by
+  letI := f.toAlgebra
+  letI := f'.toAlgebra
+  letI := g.toAlgebra
+  letI := g'.toAlgebra
+  letI := (f'.comp g).toAlgebra
+  have := RingHom.isScalarTower_toAlgebra_comp g f' _ rfl
+  have := RingHom.isScalarTower_toAlgebra_comp f g' _ fac.symm
+  have := RingHom.smulCommClass_toAlgebra g' f'
+  exact _root_.KaehlerDifferential.map_D A A' B B' b
+
+end KaehlerDifferential
+
+end CommRingCat
+
+lemma Module.isScalarTower_compHom (M : Type*) {A B : Type*}
```

Comment:
This can be an instance, and should go near `Module.compHom`.

Reply:
I wonder whether it works if `[Algebra A B]` is given and `algebraMap A B` is used instead of `f`. Can lean see through `RingHom.toAlgebra`?

Reply:
It does.